lib/ClangToAst.ml

```diff
@@ -1396,6 +1482,63 @@ let rec translate_stmt (env : env) (s : Clang.Ast.stmt_desc) : Krml.Ast.expr =
     end
   | If { cond = { desc = BinaryOperator { lhs; kind = NE; rhs }; _ }; then_branch; _ }
     when has_pointer_type lhs && is_null rhs -> translate_stmt env then_branch.desc
+
+  | If { cond; then_branch; else_branch; _ } when is_tag_check env cond ->
+      let var, variant, case_ref = deconstruct_tag_check env cond in
+
+      let lid = Helpers.assert_tlid var.typ in
+      let case, fs = match LidMap.find_opt lid !type_def_map with
+        | Some (lazy (Variant branches)) -> List.nth branches variant
+        | _ -> fatal_error "Expected a tagged union expression"
+      in
+
+      let name, case_t = match fs with
+        | [(n, (t, _))] -> n, t
+        | _ -> failwith "Tagged union case has more than one field"
+      in
+
+      let binder = Helpers.fresh_binder name case_t in
+      let pat = Krml.Ast.with_type case_t (PBound 0) in
+      (* We need to add the new binder to the environment before translating
+         the branches.
+         We also need to be able to later retrieve the content of the variant
+         constructor when accessing the corresponding field.
+         To do so, we store in the environment a variable called v!!{atom},
+         where {atom} is replaced by the binder's atom.
+         We then update the environment for `var` to store that it is a tagged
+         union in case `case`, and that the corresponding content is ``v!!{atom}`.
+         Importantly, `v!!{atom}` is not a valid variable name, and therefore does
+         not conflict with existing variables.
+      *)
+      let env_binder_name = binder.node.name ^ "!!" ^ show_atom_t binder.node.atom in
+      let new_env = add_var env (env_binder_name, case_t) in
+
+      (* We only change the state of the tagged union case to translate the if branch,
+         which is the one where we checked the tag of the variable *)
+      (* TODO: Should we sanity-check that old is None? As, if we are already in
+         a tagged union case, there is no need for rechecking the tag? *)
+      let old = !case_ref in
```

Comment:
I'm very concerned that having mutation here is going to make the code hard to reason about and maintain, see my suggestion above for the definition of variable entries in the environment

Reply:
Agreed, fixed in [01efdb5](https://github.com/AeneasVerif/scylla/pull/33/commits/01efdb58794722a7e243304ac0265ea1c202ec40)